PPC uplifting

Rot127 opened the pull request:
Supersedes https://github.com/rizinorg/rizin/pull/2497

### Overview

This PR uplifts the most common instructions of PPC32 and PPC64. Most common means almost all instructions with a occurrence of 0.001% in ~17,5 million tested (libraries for this statistic were: `libc`, `libglib`, `libcairo` and two or three more.)

The semantics of the instructions was tested against `ppc64le` and `ppc32be` binaries (see [test bins PR](https://github.com/rizinorg/rizin-testbins/pull/75)).

### Missing instructions

A handful instructions are not uplifted.

_Due to missing operand information in Capstone v4 instructions (fixed in v5):_
- `PPC_INS_SLDI` -> The immediate operand is missing in the instruction.
- `PPC_INS_ISEL` -> The `BC` operand is set to a GPR. Although it should be CR register.
- all CR logical instructions -> Capstone v4 instructions hold GPRs instead of CR registers.

_Due to missing operand information in  Capstone v5 instructions_
  - `PPC_INS_MTOCRF`, `PPC_INS_MFOCRF` -> https://github.com/capstone-engine/capstone/issues/1903

_Because they were too complex to implement_
- `PPC_INS_LWARX`: Needs some kind of "reserve address" mechanism. Also: Is disassembled to `invalid` in v4
- `PPC_INS_LMW` (0.07163% occurrence): Iterates of GPR X to 31, where x is determined during runtime -> needs a new RZIL effect.

_Because register `ca32`, `ov32` are not implemented (see TODO list below):_
   - `PPC_INS_MFXER` -> Copies a value from `xer` to a `gpr` register. QEMU and Rizin have a mismatch in register content here. The adapter can only change the value of the `xer` register. Not the `gpr` register. Hence this instruction fails in rz-tracetest.

_Others_

- Float and vector instruction are not implemented.
- Trap instructions are NOPed.

<details>
  <summary>Not implemented 0.001% instructions</summary>
  
  ```

# Trap instructions: NOPed
- [x] PPC_INS_TWLTI:        0.16662%
- [x] PPC_INS_TDLTI:        0.11114%
- [x] PPC_INS_TDNEI:        0.03006%
- [x] PPC_INS_TDEQI:        0.02348%
- [x] PPC_INS_TDLLTI:       0.02165%
- [x] PPC_INS_TDGTI:        0.02154%
- [x] PPC_INS_TWLLTI:       0.01733%
- [x] PPC_INS_TWEQI:        0.01539%
- [x] PPC_INS_TWGTI:        0.01366%
- [x] PPC_INS_TDLGTI:       0.00998%
- [x] PPC_INS_TWLGTI:       0.00752%
- [x] PPC_INS_TWNEI:        0.00668%
- [x] PPC_INS_TWUI:         0.00167%
- [x] PPC_INS_TDUI:         0.00158%

**Vector**
PPC_INS_VADDFP: 0.10065%
PPC_INS_VADDUBM: 0.06212%
PPC_INS_STVX: 0.00537%
PPC_INS_VSUBUBS: 0.00459%
PPC_INS_VCMPEQUB: 0.00391%
PPC_INS_STVEBX: 0.00141%
PPC_INS_VPKSHUS: 0.00252%
PPC_INS_XSADDDP: 0.00261%
PPC_INS_XSMADDADP: 0.00233%
PPC_INS_VPERM: 0.00369%
PPC_INS_VCMPGTSH: 0.00309%
PPC_INS_VADDUBS: 0.00204%
PPC_INS_VMSUMSHS: 0.00195%
PPC_INS_VSUBCUW: 0.00187%
PPC_INS_VSUBUWS: 0.00185%
PPC_INS_VADDSBS: 0.00185%
PPC_INS_VMHADDSHS: 0.00158%
PPC_INS_VADDCUW: 0.00158%
PPC_INS_VOR: 0.00131%
PPC_INS_VSPLTISB: 0.00130%
PPC_INS_VMINUB: 0.00127%
PPC_INS_VMSUMSHM: 0.00125%
PPC_INS_VMHRADDSHS: 0.00115%
PPC_INS_XVADDDP: 0.00130%
PPC_INS_LVX:          0.00988%
PPC_INS_VSUBSBS:      0.00577%
PPC_INS_XXSEL:        0.09343%


**Floats**
PPC_INS_FNMADD: 0.57950% -> 0x0 opcodes decode to this instruction
PPC_INS_STFDU: 0.54502%
PPC_INS_STFD: 0.35903%
PPC_INS_STFS: 0.11684%
PPC_INS_STFSU: 0.10500%
PPC_INS_FMR: 0.09640%
PPC_INS_FNMSUB: 0.09472%
PPC_INS_FNMADDS: 0.08764%
PPC_INS_FSUB: 0.07570%
PPC_INS_FCMPU: 0.06448%
PPC_INS_FMSUB: 0.05237%
PPC_INS_FCTIWZ: 0.02801%
PPC_INS_FMADD: 0.02275%
PPC_INS_FADD: 0.01949%
PPC_INS_FMUL: 0.01802%
PPC_INS_FSEL: 0.01373%
PPC_INS_FDIV: 0.01235%
PPC_INS_FRSP: 0.00821%
PPC_INS_FNEG: 0.00568%
PPC_INS_FABS: 0.00400%
PPC_INS_FMULS: 0.00320%
PPC_INS_FMADDS: 0.00300%
PPC_INS_FMSUBS: 0.00198%
PPC_INS_FCPSGN: 0.00183%
PPC_INS_FDIVS: 0.00161%
PPC_INS_FNMSUBS: 0.00146%
PPC_INS_FADDS: 0.00127%
PPC_INS_FSUBS: 0.00263%
PPC_INS_MTFSF:        0.00143%
PPC_INS_MFFS:         0.00107%
  ```
  
</details> 

### After merge

After this PR is merged issues about the missing features should be opened:

- The `dcbz` instruction zeros a cache line/block. The size of this line depends on the PPC implementation (ESIL hardcodes it to 128. RZIL to 32). It should be replaced with a plugin config option once RzArch is done.
Also, if this PR is merged before RzArch, an issue should be opened.
- All instructions which are currently not testable due to Capstone bugs in v4 and v5 should be added to an issue.
- The PPC plugin should get a `cpu=ISA300` option and implement flags `ca32` and `ov32` again. Mind to update the rz-tracetest adapter.
  - mfxer fails in rz-tracetest because of this issue if the traced binary was an ISA3 one.
- Issue about VLE. The Rizin internal `libvle` disassembles VLE instructions. Since many 32bit VLE instructions match the normal instructions by operands and exeution semantics, maybe an adapter function between `vle_t` <-> `cs_insn` could be added. This would allow to use the current code to be used by `libvle` as well.  For the 16bit instructions additional code should be written.
- Exception cases are not handled at all. Examples are: default cases of MT/MFSPR.

### Relevant other PRs this relies on:

Test binaries: https://github.com/rizinorg/rizin-testbins/pull/75
rz-tracetest adapter: https://github.com/rizinorg/rz-tracetest/pull/6
QEMU trace: https://github.com/BinaryAnalysisPlatform/qemu/pull/20

Rot127 committed:
Print warning if config for VM was NULL.

Rot127 committed:
Add warning if register is not added to VM due to overlap.

Rot127 committed:
Add several `rz_warn_if_reached` to ease debugging.

Rot127 committed:
Reformulate documentation.

Rot127 committed:
Update PPC register profile.

* Mainly adds vector and float registers. As well as some control and system registers.

Rot127 committed:
Enable to write values 4bit registers.

Rot127 committed:
PPC: Uplift most common instructions.
Uplifts:
* Branch instructions
* Fixed point instructions
* Memory load/stores

Rot127 committed:
Print warning if reserved SPR instruction is encountered.

Rot127 committed:
Fix asm tests.

Rot127 committed:
Update PC/LR addresses and add ca32 writes.

Rot127 committed:
Fix load and store tests.

* Write cache needs to get flushed so load and store tests don't share the same memory.

Rot127 committed:
Add missing T/F branch mnemonics.

Rot127 committed:
Add Move to/from CR/CR0-7

Rot127 committed:
Fix TA address calculation for branch instructions.

Rot127 committed:
Add branch tests for branch mnemonics.

Rot127 committed:
Add XNOP

Rot127 committed:
NOP cache touch instructions.

Rot127 committed:
Add undocumented ATTN instruction to the not_implemented group.

Rot127 committed:
Add isel instruction.

Rot127 committed:
Add deoxygen to macros.

Rot127 committed:
Implement CRCLR, CRSET, CROR.

* The test is marked as broken, because Capstone v4 currently decodes the wrong registers for those instructions.

Rot127 committed:
Add: CNTLZ instructions.

Rot127 committed:
Add mcrf instructions.

Rot127 committed:
Add inacive test for cmpb.

Rot127 committed:
Add Load bytes reverse instructions.

Rot127 committed:
Fix rotate tests

Rot127 committed:
Add test and add address alignment to dcbz.

Rot127 committed:
Add eqv test

Rot127 committed:
Fix typos

Rot127 committed:
COrrect DIV and MUL operations

Rot127 committed:
Update addresses after introducing another fucntion in the test binary.

Rot127 committed:
Add tests and fix High word multiplications.

Rot127 committed:
Add div tests.

Rot127 committed:
Add tests and correct MT/MFXER

Rot127 committed:
Remove register ca32, ov32

Rot127 committed:
Remove explici setting of cr register because QEMU does not do it. Otherwise we get a mismatch in the trace

Rot127 committed:
Fix BITMASK macro. The generated bitmask was incorrectly shifted to the right by one bit.

Rot127 committed:
Rework and correct CMPB algorithm. Many things were wrong. It was broken, now it isnt.

Rot127 committed:
Fix mt/mfcrf: Remove unnecessary and unitialized local var

Rot127 committed:
Simplify carry set for add and sub.

Sub instructions are exclusivly defined with addition. Hence no sub case needed.

Rot127 committed:
Unify BD and fix branch instructions. Fix: Check the single bit not the cr reg

Rot127 committed:
Document Conditional branches and replace NOPs with EMPTY

Rot127 committed:
Add mulli instruction to double word instructions.

Rot127 committed:
Fix ca set for shift instructions:

* ca value had to be determined before the shift happend. The wrong ca value was calculated if the src and target reg were the same.
* Replace NOP wit EMPTY.
* Use MSB isntead of SLT.

Rot127 committed:
Brought fixup of ADD and SUB isntructions:

The add and sub instructions had several issues which let to incorrect execution.

* The carry was incorrectly if three add operations happend (only the last add were checked, not both).
* The carry was incorrectly set if the src and target register matched.
* Same applies for the CR bit.
* It was too complex. Several local variables were introduced for this.

Rot127 committed:
Set result in local var, since it would change if src and target reg are the same.

Rot127 committed:
Fix casts for set/load sprs.

Rot127 committed:
Remove MTMSR and MFMSR since it is too complex and untestable currently.

Rot127 committed:
Use unsigned int for shift. Otherwise the 0x1c shift produces a runtime error since 0xf is int as default.

Rot127 committed:
Fix mtxer: Only write flag bits.

Rot127 committed:
Let NOT_IMPLEMENTED macro return NULL.

Rot127 committed:
Mark st[wd]cx and l[wd]cx as not implemented.

Rot127 committed:
Increase dcache_line_size to 128 bytes.

Rot127 committed:
Clearify bit meaning.

Rot127 committed:
Fix cntlz for ppc32. m was set incorrectly, since it is 0 not 32 for 32bit cpus.

Rot127 committed:
Fix isel: Use op.crx reg instead of imm.

Rot127 committed:
Unify helper function names: Prependnig `ppc_` mark as IPI

Rot127 committed:
Add more "Move to SPR" cases.

Rot127 committed:
MULLI opeartes only on double word on 64bit CPUs.

Rot127 committed:
Add documentation.

Rot127 committed:
Run clang-format

Rot127 committed:
Add untested warning to iselect.

Rot127 committed:
Cast address to fix asan build error.

Rot127 committed:
Fix ar tests:

* Most registers are now assigned the control register type and no longer show up in the ar command.

Rot127 committed:
Fix xor if dest and src registers match by saving result in local var.

Rot127 committed:
Fix BE/LE issue for Load BRX instructions.

Rot127 committed:
Update test file after grand test binary restructure.

Rot127 committed:
Fix shifts: Use only lower 6bits of n.

Rot127 committed:
User Pure local variables for ROT macros.

Rot127 committed:
Fix rldimi instructions.

* n was not inverted.
* more than 6 bits of n could be used

Rot127 committed:
Remove unnecessary macro to choose between 0 and 0 in register.

Rot127 committed:
Add asm tests.

Rot127 committed:
Update analysis test to new instruction semantics.

Rot127 committed:
Remove broken instructions and add a note about the casue and where they are implemented.

Rot127 committed:
Run clang-format

Rot127 committed:
Add isntructions previously removed by accident.

Rot127 committed:
Add 32bit emulateme tests.

Rot127 committed:
Add 64bit emulateme tests.

Rot127 committed:
Correct address after emulateme 32be was compiled with different compiler.

Rot127 committed:
Remove broken mfxer instruction from test.

Rot127 committed:
Merge branch 'dev' into ppc-rzil-uplifting

Rot127 committed:
Merge branch 'dev' into ppc-rzil-uplifting

XVilka commented on librz/analysis/arch/ppc/ppc_analysis.h:
```diff
@@ -0,0 +1,20 @@
+
+#define PPC_IN_BE_MODE (mode & CS_MODE_BIG_ENDIAN)
+// Capstone does not extract the BO or BI fields of instructions. So we do it manually.
+#define PPC_READ_BO_FIELD (((rz_read_ble32(insn->bytes, PPC_IN_BE_MODE)) & 0x03e00000) >> 21)
```
We should open a new issue for this in capstone.

Rot127 replied:
https://github.com/capstone-engine/capstone/issues/1909

XVilka commented on librz/analysis/arch/ppc/ppc_il.c:
```diff
@@ -0,0 +1,882 @@
+RZ_IPI ut32 ppc_fmx_to_mask(const ut8 fmx) {
+	ut32 x = 0xf;
+	return (
+		(fmx & 0x80 ? x << 0x1c : 0) |
```
In general, for shifter values, it's better to use decimal numbers rather than hexadecimal.

XVilka commented on librz/analysis/arch/ppc/ppc_il.c:
```diff
@@ -0,0 +1,882 @@
+		(fmx & 0x01 ? x : 0));
+}
+
+// This function is a prerequisite for a broken insruciton implementation.
```
A typo in "instruction"

XVilka commented on librz/analysis/arch/ppc/ppc_il.c:
```diff
@@ -0,0 +1,882 @@
+// It has been already implemented.
+// For the implementation see: https://github.com/Rot127/rizin/tree/ppc-rzil-broken-insn-impl
+//
+// RZ_IPI ut8 ppc_translate_cs_cr_flag(const char *flag);
```
Is it really necessary? Maybe remove it?

Rot127 replied:
The hint to the implemented functions? No. I would add it to the issue then?

XVilka commented on librz/analysis/arch/ppc/ppc_il.c:
```diff
@@ -0,0 +1,882 @@
+ * \return bool True if the instruction operates on double words. False otherwise;
+ */
+RZ_IPI bool ppc_is_mul_div_d(const ut32 id, const cs_mode mode) {
+	return id == PPC_INS_MULHD || id == PPC_INS_MULLD || id == PPC_INS_MULHDU || id == PPC_INS_DIVD || id == PPC_INS_DIVDU || ((id == PPC_INS_MULLI) && IN_64BIT_MODE);
```
Split the line

XVilka commented on librz/analysis/arch/ppc/ppc_il.h:
```diff
@@ -0,0 +1,183 @@
+RZ_IPI ut32 ppc_fmx_to_mask(const ut8 fmx);
+RZ_IPI RZ_OWN RzILOpPure *ppc_get_cr(const ut8 x);
+RZ_IPI const char *ppc_get_cr_name(const ut8 x);
+// Prerequisite for broken instruction implementation.
```
What "broken instruction implementation" means?

XVilka commented on librz/analysis/arch/ppc/ppc_il_flag_ops.c:
```diff
@@ -0,0 +1,64 @@
+ * \param b Value b.
+ * \param c Value c. Is optional and can be NULL
+ * \param mode Capstone mode.
+ * \return RZ_OWN* Effect which sets the carry bits.
```
Shouldn't be `RzILOpEffect` instead?

XVilka commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -0,0 +1,1597 @@
+		NOT_IMPLEMENTED;
+	case PPC_INS_DCBZ:;
+		// Determine log_2(dcache_line_size)
+		// See: https://graphics.stanford.edu/~seander/bithacks.html#IntegerLogObvious
```
Maybe extract this into a small static function? It will be inlined

XVilka commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -0,0 +1,1597 @@
+		//
+		// Bug:
+		// Capstone v5 does not store the fxm value in the operands.
+		// See: https://github.com/capstone-engine/capstone/issues/1903
```
Could you please list all capstone bugs you met in the PR description?

Rot127 replied:
Updated the description.

XVilka commented on librz/reg/rvalue.c:
```diff
@@ -229,6 +229,20 @@ RZ_API bool rz_reg_set_value(RzReg *reg, RzRegItem *item, ut64 value) {
 	case 8:
 		rz_write_ble8(src, (ut8)(value & UT8_MAX));
 		break;
+	case 4:;
```
A typo

Rot127 replied:
The next line is a statement but after a `case` must follow an expression. The single `;` is there as a single expression. Otherwise it will not compile.

XVilka replied:
A better way is to just add `{` and `}` instead, no?

Rot127 replied:
Yeah, you are right. Looks indeed better.

XVilka reviewed (changes requested):
Solid work!

Rot127 committed:
Use decimal values for shifting.

Rot127 committed:
Remove hints for broken but implemented instructions. They will be listed in an issue.

Rot127 committed:
Formatting and typo

Rot127 committed:
Determine lg(v) in inline function.

Rot127 committed:
Put code which starts with a statement after a `case` into brackets to avoid `:;` usage.

Rot127 commented:
Just discovered two mismatches in the trace of `emulateme`. Please let me check them before merge.

Rot127 commented:
My bad. Just used an old trace which was not correct. Everything works fine.

wargio commented:
@Rot127 is VLE implemented?

Rot127 commented:
> is VLE implemented?

No. Currently this runs exclusively on Capstone and it still misses VLE if I'm not mistaken.

wargio commented:
> > is VLE implemented?
> 
> No. Currently this runs exclusively on Capstone and it still misses VLE if I'm not mistaken.

Yes capstone has no VLE, but i wrote libvle which can be used for this too, do not worry, it is not needed now.

Rot127 commented:
> > > is VLE implemented?
> > 
> > 
> > No. Currently this runs exclusively on Capstone and it still misses VLE if I'm not mistaken.
> 
> Yes capstone has no VLE, but i wrote libvle which can be used for this too, do not worry, it is not needed now.

Ah yes. I have seen this of cause. This should really be added a bit later. Naturally I make heavy use of `cs_insn.detail.ppc` (struct which holds all the operand information). So for example we couldn't pass the `e_stb` instruction to the `stb` uplifting code without problems.
But I add it to the "issue after merge" list. Maybe some kind of adapter function between `vle_t` <-> `cs_insn` could be added (for the simple cases). And additional code for all the others.

wargio commented on librz/analysis/arch/ppc/ppc_il.c:
```diff
@@ -0,0 +1,877 @@
+	RzILOpPure *field_bit;
+	RzILOpPure *cr_field;
+	if (pos < 36) {
+		field_bit = SHIFTR0(UN(4, 0b1000), UN(4, pos - 32));
+		cr_field = VARG("cr0");
+	} else if (pos < 40) {
+		field_bit = SHIFTR0(UN(4, 0b1000), UN(4, pos - 36));
+		cr_field = VARG("cr1");
+	} else if (pos < 44) {
+		field_bit = SHIFTR0(UN(4, 0b1000), UN(4, pos - 40));
+		cr_field = VARG("cr2");
+	} else if (pos < 48) {
+		field_bit = SHIFTR0(UN(4, 0b1000), UN(4, pos - 44));
+		cr_field = VARG("cr3");
+	} else if (pos < 52) {
+		field_bit = SHIFTR0(UN(4, 0b1000), UN(4, pos - 48));
+		cr_field = VARG("cr4");
+	} else if (pos < 56) {
+		field_bit = SHIFTR0(UN(4, 0b1000), UN(4, pos - 52));
+		cr_field = VARG("cr5");
+	} else if (pos < 60) {
+		field_bit = SHIFTR0(UN(4, 0b1000), UN(4, pos - 56));
+		cr_field = VARG("cr6");
+	} else {
+		field_bit = SHIFTR0(UN(4, 0b1000), UN(4, pos - 60));
+		cr_field = VARG("cr7");
```
couldn't this `field_bit` field be pre-calculated in C instead of doing it in IL?

wargio replied:
why this was not simplified?

wargio commented on librz/analysis/arch/ppc/ppc_il.c:
```diff
@@ -0,0 +1,877 @@
+ * \return RzILOpEffect* Sequence of effects to sync the CR/CRx registers.
+ */
+RZ_IPI RZ_OWN RzILOpEffect *ppc_sync_crx_cr(const bool crx_to_cr, const ut32 cr_mask) {
+	RzILOpEffect *sync;
```
please rename this variable to `effect`, since sync is kinda a special word.

wargio commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -0,0 +1,1587 @@
+ * \param v Value to determine log_2(v) for.
+ * \return ut32 lg(v)
+ */
+static inline ut32 ppc_log_2(const ut32 v) {
+	ut32 x = v;
+	ut32 r = 0; // r will be lg(v)
+	while (x >>= 1) {
+		r++;
+	}
+	return r;
+}
```
Please add a comment above saying that could be simplified using some fast log2 approximation algorithm.

wargio commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -0,0 +1,1587 @@
+	switch (id) {
+	default:
+		NOT_IMPLEMENTED;
+	// ???
```
```suggestion
```

wargio commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -0,0 +1,1587 @@
+		ut32 spr = INSOP(1).imm;
+		switch (spr) {
+		default:
+			if (spr & 1) {
+				// TODO
+				// Invoke system privileged instruction error handler
+			} else {
+				// TODO
+				// Invoke illegal instruction handler
+			}
```
these are empty, so probably is better to just keep the 2 todos.

wargio commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -0,0 +1,1587 @@
+		NOT_IMPLEMENTED;
+	case PPC_INS_ROTLW:
+	case PPC_INS_ROTLWI:
+	case PPC_INS_RLWIMI:
```
This can be heavily simplified: `dst = (dst & ~mask) | (rotl32(src, sh) & mask)`
if `mask` is 0, then is just `dst &= ~0`
if `~mask` is 0, then is just `rotl32(src, sh)`
if `sh` is 0, then is just `src`.

wargio replied:
if you want to implement this, i suggest to test first the mask then the shift.

Rot127 replied:
I will look into it. But the first case (mask = 0) is not possible by the ISA. The mask algorithm guaranties to set at least one bit.

wargio commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -0,0 +1,1587 @@
+	case PPC_INS_ROTLW:
+	case PPC_INS_ROTLWI:
+	case PPC_INS_RLWIMI:
+	case PPC_INS_RLWINM:
```
also this one can be simplified in the same way as above.

wargio commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -0,0 +1,1587 @@
+		break;
+	case PPC_INS_ROTLD:
+	case PPC_INS_ROTLDI:
+	case PPC_INS_RLDCL:
+	case PPC_INS_RLDCR:
+	case PPC_INS_RLDIC:
+	case PPC_INS_RLDICL:
+	case PPC_INS_RLDICR:
+	case PPC_INS_RLDIMI:
```
These can be simplified based if their mask and shift is 0.

wargio reviewed (changes requested):
Incredible job. well done, i added some comments

Rot127 committed:
Remove empty conditionals and leave TODO message.

Rot127 committed:
Replace obvious lg algorithm with faster version.

Rot127 committed:
Rename sync -> effect and remove useless comment.

Rot127 committed:
Calculate CR bit in C not in the VM.

Rot127 committed:
Check if `~mask = 0` and skip mask calculation if yes.

Rot127 committed:
Fix asm tests for last two commits.

XVilka reviewed (approved):
Please open an issue with all points that are deferred to future work to not forget about them.

Rot127 committed:
Check for `sh == 0` and skip rotations where possible.

Rot127 commented:
@wargio Implemented the simplification for `~mask = 0` and some `shift = 0` cases. Not all though:

Currently instructions which take the shift amount from an immediate and a register are grouped together. So no pure C operations possible without refactoring the whole function. But the bitvector shift returns anyways if `n == 0`. So adding `EQ(n, 0)` would add bloat, not reduce it.

The ROT32 instructions can not be skipped, since in 64bit mode the 32bit values are concatenated.
Added the simplifications for the ROT64 cases.

Do you think refactoring the rotate/shift code is worth an issue? I don't like it and frankly it had also the most bugs.

XVilka commented:
@Rot127 @wargio I think if such simplification is planned, it's better to do it in a separate PR; also easier to review. If no major objections to this - I suggest merging this as is and continuing to improve separately. Moreover, it will benefit from stuff like CodeQL, Coverity, fuzzing tests, etc that happen on the `dev` branch periodically.

wargio commented on librz/analysis/arch/ppc/ppc_il_ops.c:
```diff
@@ -1069,7 +1069,7 @@ static RzILOpEffect *shift_and_rotate(RZ_BORROW csh handle, RZ_BORROW cs_insn *i
 		}
 		n = LOGAND(U8(0x3f), n);
-		r = ROTL64(VARG(rS), n);
+		r = LET("n", n, ITE(EQ(VARLP("n"), U8(0)), VARG(rS), ROTL64(VARG(rS), VARLP("n"))));
```
in this case i prefer the previous one.

Rot127 committed:
Revert `n == 0` check for rotate instructions.

wargio commented:
i will optimize those checks later, in another PR

thestr4ng3r reviewed (changes requested):
A few valid instructions are still missing tests:
https://app.codecov.io/gh/rizinorg/rizin/blob/ppc-rzil-uplifting-cov/librz/analysis/arch/ppc/ppc_il_ops.c
https://app.codecov.io/gh/rizinorg/rizin/blob/ppc-rzil-uplifting-cov/librz/analysis/arch/ppc/ppc_il.c

Please add asm tests for those, and perhaps also some for ppc_32.

Rot127 committed:
Remove unused code.

Rot127 committed:
Remove `la` instruction.

`la` is a mnemonic for `addi`. I could not find a single occurrence in any of the test binaries of it.
The assembler always resolves the `la` mnemonic into `addi`.
I couldn't find a documented bit difference between `la` and `addi`. Because the instruction can not be tested it is removed.

Rot127 committed:
Remove SPR instructions which are not supported by QEMU or not traced yet.

For most set/read SPR instructions QEMU segfaults.
Those which don't cause segfaults are not traced in QEMU currently.
Because they are barely used anyways I removed them (instead if imeplenting the tracing in QEMU which would have cost too much time for me currently).

In case of SPR 1 (xer), 8 (lr) and 9 (ctr) the assembler resolves them to their mnemonics (mtxer, mtlr etc.). This means the code here is never reached.
To test the get_xer code MFXER was added again. The rz-tracetests will fail for this instructions (due to missing ca32, ov32). But this case is covert in an issue.

Rot127 committed:
Add little endian tests for load bytes reverse instructions.

Rot127 committed:
Add test which covers cr5 check.

This should cover the uncovered branch in `get_cr_bit`

Rot127 commented:
@thestr4ng3r Regarding `la`: I thought about assembling a `la` instruction manually and add it to the asm tests but decided against it eventually. Because it seems to be resolved by the assembler anyways and I couldn't find a documented bit difference between `addi` <-> `la`.

wargio commented:
> la
![image](https://user-images.githubusercontent.com/561184/184011346-4d7d2f7e-bc5c-4347-9264-1dd772a7a4f6.png)

`la` is `addi`

XVilka commented:
Since @Rot127 addressed your comments @thestr4ng3r, I am merging. If you have any additional feedback, please open a new issue, ok?
